datefudge.man
.\" vim:ft=nroff
.TH DATEFUDGE "1" "August 1th, 2019" "datefudge @VERSION@" Debian
.SH NAME
datefudge \- pretend the system time is different
.SH SYNOPSIS
\fBdatefudge\fR [\fB\-s\fR|\fB\-\-static\fR] [\fB\-l\fR|\fB\-\-add\-ld\-preload\fR \fIlib\fR]
[\fB\-f\fR|\fB\-\-from\-file\fR \fIdate_file_path\fR] [\fB\-c\fR|\fB\-\-cache\-file\-date\fR \fIms\fR] [\fIdate\fR] program\fR [\fIarguments\fR ...]
.SH DESCRIPTION
.PP
\fBdatefudge\fR is a small utility that pretends that the system time is different
by pre-loading a small library which modifies the
\fBtime\fR(2), \fBgettimeofday\fR(2) and \fBclock_gettime\fR(2) system calls.
.SH DATE FORMAT
The \fIat_date\fP argument can be given in any format accepted by the \fBdate\fP(1)
program, for example "2007\-04\-01 12:21" or "yesterday", or "next Friday".
.PP
The date passed to program inside \fIdate_file_path\fP supports full or partial
date-time string to the seconds level, for example "2007\-04\-01 12:21:12",
"2007\-04\-01 12", "2007\-04\-01" or "2007".
File is checked for updates at most each 100 milliseconds (this value can be
overriden by arg 'ms').
.SH OPTIONS
.TP
\fB\-\-static\fR, \fB\-s\fR
Mark the date as a `static' one.  The above mentioned system calls will always return
the date given in the \fIat_date\fP argument, regardless of time passing.  See EXAMPLES
below.
.TP
\fB\-\-add\-ld\-preload\fR \fIlib\fR, \fB\-l\fR \fIlib\fR
Prepend \fIlib\fR to \fILD_PRELOAD\fR environment variable before executing given
\fIprogram\fR.  This option might be useful for example to inject sanitizer
libraries (e.g. from Address Sanitizer) to the list of preloaded libraries before the
internal \fBdatefudge\fR's library.  See EXAMPLES below.
.TP
\fB\-\-help\fR, \fB\-h\fR
Print short usage information and exit.
.TP
\fB\-\-version\fR, \fB\-v\fR
Print version information and exit.
.SH EXAMPLES
.SS Basic examples:
.nf
\fI$ datefudge "2007\-04\-01 10:23" date \-R\fR
Sun, 01 Apr 2007 10:23:00 +0200

\fI$ datefudge "1 May 2007" date \-R\fR
Tue, 01 May 2007 00:00:00 +0200

\fI$ datefudge "2 weeks ago" date \-R\fR
Wed, 16 Jan 2008 13:32:12 +0100
.fi

.SS Date-time from file examples:
date.txt contains "2007\-04\-01 10:23:45".
.nf
\fI$ datefudge -f date.txt date \-R\fR
Sun, 01 Apr 2007 10:23:45 +0200

.fi
date.txt contains "2007\-04\-01 10".
.nf
\fI$ datefudge -f date.txt date \-R\fR
Sun, 01 Apr 2007 10:00:00 +0200

.fi
date.txt contains "2007\-04\-01".
.nf
\fI$ datefudge -f date.txt date \-R\fR
Sun, 01 Apr 2007 00:00:00 +0200
.fi

.SS Non-static vs. static example:
.nf
\fI$ datefudge "2007\-04\-01 10:23" sh \-c "\fBsleep 3\fI; date \-R"\fR
Sun, 01 Apr 2007 10:23:\fB03\fR +0200

\fI$ datefudge \fB\-\-static\fI "2007\-04\-01 10:23" sh \-c "\fBsleep 3\fI; date \-R"\fR
Sun, 01 Apr 2007 10:23:\fB00\fR +0200
.fi
.SS Prepending LD_PRELOAD example:
.nf
\fI$ datefudge \fB\-\-add\-ld\-preload /path/to/libclang_rt.asan\-x86_64.so\fI "2007\-04\-01 10:23" command\fR

If the \fIcommand\fR was compiled by 'clang \-fsanitize=address \-shared\-libasan', then the above
command might display some errors, for example:

==7625==ERROR: AddressSanitizer: heap\-use\-after\-free on address 0x614000000044 at pc ....

Without the \fB\-\-add\-ld\-preload\fR argument, this would not work at all.
.fi
.SH AUTHOR
Written by Matthias Urlichs <smurf@noris.de>.  Modified by
Robert Luberda <robert@debian.org>.
.SH BUGS
There is no attempt to make this change undetectable by the program.
In particular, file modification times are \fBnot\fR modified.
.PP
On systems using 32-bit representation of time, \fBdatefudge\fR is affected by the year
2038 problem, which might cause dates to be wrapped around, for instance:
.nf
\fI$ TZ=UTC datefudge "\fB2038\fR\fI\-01\-19 03:14:07" sh \-c "sleep 1; date \-R"\fR
Fri Dec 13 20:45:53 UTC \fB1901\fR
.fi
.SH COPYRIGHT
Copyright \(co 2003 by Matthias Urlichs.
.br
Copyright \(co 2008-2019 by Robert Luberda.
.PP
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  You may redistribute copies of datefudge under the
terms of the GNU General Public License.
.br
For more information about these matters, see the file named COPYING.
.SH "SEE ALSO"
\fBdate\fR(1), \fBld.so\fR(1), \fBtime\fR(2), \fBgettimeofday\fR(2), \fBclock_gettime\fR(2)


debian/changelog
datefudge (1.25) unstable; urgency=medium

  [ Vedran Vidović ]
  * Added option to read date-time from a file - this enables us to change
    date & time during a lifetime of some long-running process.

 -- Vedran Vidović <vvidovic@gmail.com>  Wed, 21 Oct 2020 10:24:26 +0200

datefudge (1.24) unstable; urgency=medium

  [ Debian Janitor ]
  * Update standards version to 4.4.1, no changes needed.
  * Set debhelper-compat version in Build-Depends.
  * Fix day-of-week for changelog entries 1.12, 1.11, 1.10, 1.9, 1.8,
    1.7, 1.6, 1.5, 1.4, 1.3, 1.2, 1.1, 1.0.

  [ Robert Luberda ]
  * Fix FTBFS on hurd-i386: Necessary functions disabled (closes: #945599).
    Thanks to Paul Sonnenschein for the patch.
  * Fix FTBFS with glibc 2.31 (conflicting gettimeofday prototype) (closes:
    #964227). Thanks to Aurelien Jarno for the patch.
  * Bump debhelper's compact level to 13.
  * Standards-Version: 4.5.0.

 -- Robert Luberda <robert@debian.org>  Tue, 14 Jul 2020 11:11:48 +0200

datefudge (1.23) unstable; urgency=medium

  [ Ondřej Nový ]
  * d/control: Set Vcs-* to salsa.debian.org

  [ Robert Luberda ]
  * Makefile: No longer require root privileges for installation;
    set `Rules-Requires-Root' to `no' in debian/control.
  * Build with debhelper v12.
  * datefudge.sh: introduce '--add-ld-preload|-l arg' option to prepend
    given arg to LD_LPRELOAD variable, what might be useful for Address
    Sanitizer and other santizers (closes: #873577).
  * datefudge.man: document the above option.
  * Add simple tests for autopkgtest.
  * Standards-Version: 4.4.0.

 -- Robert Luberda <robert@debian.org>  Fri, 02 Aug 2019 20:09:51 +0200

datefudge (1.22) unstable; urgency=medium

  * datefudge.c: Apply the following fixes from Thorsten Glaser
    (closes: #853724):
    + Operate more efficiently, especially if fudging to the epoch.
    + Fix running on 32-bit systems with 64-bit time_t (such as x32).
  * datefudge.man: Document that datefudge might be affected by the
    Y2K38 problem on some systems.
  * Use https in Vcs-Git and switch Vcs-Browser to cgit.
  * Standards-Version: 3.9.6 (no changes).

 -- Robert Luberda <robert@debian.org>  Sat, 04 Feb 2017 22:59:11 +0100

datefudge (1.21) unstable; urgency=medium

  * datefudge.sh: don't fail when its date argument happens to be equal
    to the current time (i.e. make `datefudge now some_command' work).
  * Enable all hardening options in debian/rules.
  * Update man page to mention that the date can be given in any format
    accepted by date(1) command.
  * Switch debian/copyright to the DEP-5 format.
  * Use https in Vcs-Browser field (lintian).
  * Standards-Version: 3.9.6 (no changes).

 -- Robert Luberda <robert@debian.org>  Sat, 13 Feb 2016 15:19:41 +0100

datefudge (1.20) unstable; urgency=medium

  * datefudge.c: don't override time(2) on GNU/Hurd, as it seems to
    internally call gettimeofday(2), which is overridden. This fixes
    FTBFS on that platform.

 -- Robert Luberda <robert@debian.org>  Mon, 16 Dec 2013 23:32:47 +0100

datefudge (1.19) unstable; urgency=medium

  * Use `*-*' as multi-arch pattern to make sure the datefudge script is the
    same on all architectures, including 'hurd-i386' (closes: #730766).
  * Install datefudge.so into /usr/lib<arch>/datefudge instead of using
    an extra private dir /usr/lib/datefudge/<arch> (closes: #711892).
  * Cleanup makefiles a bit and add support for cross-compiling.
  * Standards-Version: 3.9.5 (no changes).

 -- Robert Luberda <robert@debian.org>  Sat, 14 Dec 2013 21:06:22 +0100

datefudge (1.18) unstable; urgency=low

  * Introduce multi-arch support which allows one to use datefudge with 32-bit
    commands on 64-bit system (providing that both 32-bit and 64-bit versions
    of this package are installed on system) (LP: #1103281):
    + install datefudge.so library into /usr/lib/datefudge/<arch-tripplet>;
    + add each /u/l/d/<arch-tripplet> to LD_LIBRARY_PATH in datefudge scipt;
    + mark binary package as `Multi-Arch: same'.
  * Use debhelper v9.
  * debian/control:
    + fix Vcs-Browser field;
    + Standards-Version: 3.9.4 (no changes).

 -- Robert Luberda <robert@debian.org>  Sat, 11 May 2013 19:02:11 +0200

datefudge (1.17) unstable; urgency=low

  * datefudge.{c,man}: Add the `--static' option for static dates.
  * debian/control:
    + Standards-Version: 3.9.2 (no changes);
    + add VCS fields.

 -- Robert Luberda <robert@debian.org>  Thu, 23 Jun 2011 19:38:38 +0200

datefudge (1.16) unstable; urgency=low

  * datefudge.c: don't override clock_gettime(2) on GNU/Hurd - it seems
    it internally calls gettimeofday(2), which is overridden. This fixes
    FTBFS due to test failure on that platform.
  * Makefile: add test for `perl localtime()'.

 -- Robert Luberda <robert@debian.org>  Mon, 14 Feb 2011 23:03:20 +0100

datefudge (1.15) unstable; urgency=low

  * Use the `3.0 (native)' source format.
  * Switch to debhelper v8.
  * Standards-Version: 3.9.1 (no changes).
  * debian/rules:
    + provide build-arch and build-indep targets;
    + call dpkg-buildflags to set CFLAGS and LDFLAGS;
    + simplify the file.
  * Reword the restriction in README (closes: #416174).
  * Update debian/copyright a bit.
  * Makefile: add a simple test case.

 -- Robert Luberda <robert@debian.org>  Sun, 13 Feb 2011 13:28:08 +0100

datefudge (1.14) unstable; urgency=low

  * New maintainer (closes: #429467).
  * Standards-Version: 3.7.3 (no changes).
  * Use debhelper v7 and its minimised rules file.
  * datefudge.c: override clock_gettime(2) called with CLOCK_REALTIME
    argument to make it possible to fake `date' command (closes: #416175).
  * Update man page, add a simple example.

 -- Robert Luberda <robert@debian.org>  Sat, 10 May 2008 08:21:02 +0200

datefudge (1.13) unstable; urgency=low

  * QA upload.
  * Package is orphaned (#429467); set maintainer to Debian QA Group.
  * Switch to debhelper.
  * Move datefudge.so to /usr/lib/datefudge as it's not a public library.

datefudge.sh
#!/bin/sh
# vim:noet:ts=2:sw=2:et
# Fake the system time

dat=""
static=0
date_file_path=""
date_file_cache_ms="100"
ld_preload_arg=""

usage()
{
  if [ "$1" ]; then
    exec >&2
    echo "$0: $1"
    echo
  fi
  echo "Usage: $0 [-s|--static] [-l|--add-ld-preload lib] [-f|--from-file date_file_path] [-c|--cache-file-date ms] [date] program args..."
  echo
  echo "Run 'program' with 'args'."
  echo "The program will believe that the current time is 'date' (when"
  echo "date_file_path is not used) or full/partial date-time set in"
  echo "date_file_path (yyyy-MM-dd HH:mm:ss)."
  echo
  echo "When reading date/time from a file, it is checked for updates at most"
  echo "each 100 milliseconds (this value can be overriden by param 'ms')."
  echo
  [ "$1" ] && exit 1 || exit 0
}

while [ "$1" ] && [ -z "$dat" ]; do
  case "$1" in
    -s|--static)
      static=1
      ;;
    -l|--add-ld-preload)
      [ "$2" ] || usage "Missing argument for the '$1' option"
      ld_preload_arg="$2"
      shift;
      ;;
    -f|--from-file)
      [ "$2" ] || usage "Missing argument for the '$1' option"
      date_file_path="$2"
      shift;
      ;;
    -c|--cache-file-date)
      [ "$2" ] || usage "Missing argument for the '$1' option"
      date_file_cache_ms="$2"
      shift;
      ;;
    -v|--version)
      echo "$0: Version @VERSION@"
      echo ""
      echo "For usage information, use '$0 --help'."
      exit 0
      ;;
    -h|-?|--help)
      usage
      ;;
    -*)
      usage "Invalid option '$1'"
      ;;
    *)
      # If we have date file path param we expect only the program param
      # otherwise we expect the date param & the program param.
      if [ -z "$date_file_path" ]; then
        [ "$2" ] || usage "Missing 'program' argument"
        dat="$1"
      else
        [ "$1" ] || usage "Missing 'program' argument"
        break
      fi
      ;;
  esac
  shift
done

# One of params must be set: date or date_file_path
if [ -n "$dat" ]; then
  # Assume that 'date' already printed an error message
  sec1=$(date -d "$dat" '+%s')
  [ $? -eq 0 ] || exit 1

  sec2=$(expr $(date '+%s') - $sec1)
  # According to its documentation expr returns exit status 1,
  # when the expression evaluates to 0
  [ $? -le 1 ] || exit 1
elif [ -z "$date_file_path" ]; then
  usage "Missing 'date' or 'date_file_path' argument"
fi

add_ld_preload()
{
  export LD_PRELOAD="${1}${LD_PRELOAD:+:}${LD_PRELOAD}"
}

add_ld_library_path()
{
  export LD_LIBRARY_PATH="${1}${LD_LIBRARY_PATH:+:}${LD_LIBRARY_PATH}"
}

set_ld_environment()
{
  lib="datefudge.so"
  libpath="@LIBDIR@"
  set --  "$libpath"@MULTIARCH_PATTERN@"/datefudge/$lib"
  if [ ! -e "$1" ]; then
    echo "Cannot find $lib in $libpath" >&2
    [ -z "$POSH_VERSION" ] || echo "You might have just encountered posh bug#636601, please try using another shell." >&2
    exit 1;
  fi
  for path in "$@"; do
    add_ld_library_path "${path%/*}"
  done
  add_ld_preload "$lib"
  [ -z "$ld_preload_arg" ] || add_ld_preload "$ld_preload_arg"
}

set_datefudge_vars()
{
  if [ "$static" = "1" ]; then
    export DATEFUDGE=$sec1
    export DATEFUDGE_DOSTATIC="1"
  else
    export DATEFUDGE=$sec2
    unset  DATEFUDGE_DOSTATIC
  fi

  if [ -n "$date_file_path" ]; then
    export DATEFUDGE_FILE=$date_file_path
    export DATEFUDGE_FILE_CACHE_MS=$date_file_cache_ms
    unset  DATEFUDGE
  fi
}


set_ld_environment
set_datefudge_vars

exec "$@"
